### ISSUE TITLE
DPM / Widget-DPM Status Tracker

### ISSUE THREAD

**Issue by candiceT233** [pnnl/DataFlowDrs#1]
This is an status summary for replacing the old standalone DPM workflow with Widget-DPM in the paper/evaluation workflow.

## Repositories

- Widget: <https://github.com/candiceT233/widget-v1>
- Orchestrator, Hasan fork: <https://github.com/candiceT233/agentic-workflow-orchestrator>
- Old DPM: <https://github.com/candiceT233/dpm>

### Old DPM Provenance Identified

Status: completed.

The old DPM repo corresponds to the paper-era implementation and remains useful
for provenance, historical result comparison, and plotting/result checks. It is
not the preferred place for new Widget-DPM implementation.

Important details:

- Old DPM uses exhaustive storage-resource profiling/training.
- Old DPM workflow DAG construction has unnecessary repeated work.
- Old DPM paper selections checked so far are consistent with the intended
  storage-parallelism claims:
  - 1000Genome: `TMPFS 10n`.
  - PyFLEXTRKR: `BeeGFS 8n` after accounting for data movement.
  - Montage legacy result: paper/source and local measured logs support
    `BeeGFS 16n`, but Montage is now planned to be replaced by SRAsearch.

### Widget-DPM Core Replacement

Status: completed for current paper-replacement use.

Widget-DPM now provides the new implementation path for DPM prediction/export.

Completed details:

- Adaptive/data-efficient storage profiling support exists in Widget planning
  scripts and docs. The current target is about 80% fewer benchmarks than the
  old exhaustive profiler, pending final experiment wording.
- Workflow DAG construction is handled through Widget trace parsing and DAG
  caching, which removes repeated old-DPM DAG-building work and improves
  practical runtime.
- DPM regression/prediction backends include interpolation and random forest.
  Current preference is interpolation because it avoids RF training overhead and
  worked better for the recent verification path.
- Full candidate-plan DPM export is implemented via `dump_candidate_plan_dpm`.
- Repeated task-pair/prediction structures can be compressed in export using
  `compression: "task_pair"` or `compression: "prediction_signature"`.
- Missing movement profiles are now represented as not-applicable candidates
  rather than silently using an invalid proxy.
- Data movement logic now handles:
  - shared same-storage access such as `beegfs_x_beegfs`;
  - local same-storage movement only when producer/consumer effective node
    footprints differ;
  - 1000Genome chromosome-local no-movement internal edges for same-storage
    local candidates at the trace-locality scale.

### Widget-DPM Tests

Status: completed for focused coverage.

Focused tests passed:

```text
54 passed
```

Covered scenarios include:

- interpolation fallback for lower-dimensional benchmark slices;
- candidate-plan export validation;
- missing data-movement profile handling;
- 1000Genome local no-movement behavior;
- PyFLEXTRKR local movement behavior when effective producer/consumer local
  footprints differ;
- DPM-MT contention post-processing.

### Widget API / Agent Interaction

Status: completed for the current interface.

Widget exposes the DPM workflow through MCP tools and is agent-friendly.

Relevant tools:

- `dump_candidate_plan_dpm`
- `prepare_and_dump_dpm`
- `predict_dpm_space`
- `analyze_workflow_dpm`
- `get_workflow_node_options`
- visualization/profile tools such as Sankey and benchmark plotting helpers

Output is organized as dictionary/JSON artifacts rather than ad hoc notebook or
plot-script state.

### Documentation Refresh

Status: completed for first pass.

Updated and added docs:

- `README.md`: command-first usage, Widget component overview, DPM and DPM-MT
  entry points.
- `docs/DPM_EXPORTS.md`: full candidate-plan artifact usage and schema notes.
- `docs/WIDGET_COMPONENT_MAP.md`: how Widget components connect.
- `docs/MULTI_TENANT_DPM_MODEL.md`: DPM-MT prototype model.
- `docs/IN_PROGRESS_ITEMS.md`: prototype/future work tracker.
- `docs/PAPER_WORKFLOW_REPLACEMENT_NOTES.md`: Montage caveat and SRAsearch
  starting point.

### DaYu Integration

Status: partially completed.

What is completed:

- DaYu adapter files exist in Widget.
- MCP tools are registered for:
  - `run_dayu_profiling`
  - `dayu_vol_sankey`
  - `dayu_vfd_sankey`
  - `dayu_combined_sankey`
- Documentation exists in `docs/DAYU_INTEGRATION.md`.

Important caveat:

- DaYu has not yet had a full end-to-end validation test in the current
  Widget-DPM replacement workflow.
- `src/dfl_mcp/verification.py` still has stub/light validation for DaYu.

Short answer to "is DaYu integration done?":

DaYu is integrated at the adapter/tool/documentation level, including HDF5 VOL
and VFD Sankey analysis paths, but it should not be called fully validated until
an end-to-end HDF5 workflow test passes and verification is upgraded.

### DPM-MT / Background Traffic Prototype

Status: prototype completed, not calibrated.

What exists:

- Separate extension package: `src/dfl_mcp/dpm_mt`.
- Script: `scripts/apply_dpm_mt_contention.py`.
- Documentation: `docs/MULTI_TENANT_DPM_MODEL.md`.
- Roadmap notes: `docs/IN_PROGRESS_ITEMS.md`.

Current usage:

```bash
.venv/bin/python scripts/apply_dpm_mt_contention.py \
  --input output/logs/results/candidate_plan_dpm.json \
  --output output/logs/results/candidate_plan_dpm_mt.json \
  --background beegfs:write:0.4
```

Short answer to "where to look for DPM in context of background traffic?":

Look at `src/dfl_mcp/dpm_mt/` and `docs/MULTI_TENANT_DPM_MODEL.md`. It is a
post-processing framework for full candidate-plan DPM artifacts. It models
assumed background occupancy with a capacity-share slowdown. It is not yet a
validated multi-tenant storage predictor.

### README / HOWTO Expansion

Status: todo, first pass complete.

Need to add more user-facing HOWTO topics. Prefer command-first docs followed
by parameter explanations.

Suggested topics:

- Generate a full DPM artifact for a workflow.
- Apply DPM-MT background contention.
- Run adaptive storage profiling.
- Build and run DataLife profiling.
- Build and run DaYu profiling.
- Generate Sankey/data lifecycle plots.
- Compare Widget-DPM output against old DPM paper selections.
- Prepare a workflow trace directory for Widget ingestion.
- Use Widget-DPM from an agent/MCP client.

### Experiment Validation Data Archiving

Status: todo.

Need to collect validation data into:

```text
/qfs/projects/oddite/dpm
```

Requested actions:

- Compress for archival.
- Label training/profile data so old exhaustive training data can be
  deprecated cleanly.
- Keep raw evaluation data in the DPM/provenance area, not mixed into Widget
  source.

### SRAsearch Replacement Workflow

Status: todo, local starting point found.

Local starting point:

```text
/qfs/projects/datamesh/tang584/fastflow_scripts/srasearch
```

Known workflow skeleton:

```text
SRA accession
  -> fasterq-dump
  -> paired FASTQ
  -> bowtie2
  -> samtools view
  -> samtools sort
  -> BAM
  -> samtools index
  -> BAI
```

Still needed:

- Create a reproducible workflow driver or scheduler workflow.
- Verify tool paths and environment.
- Run DataLife/block tracing.
- Generate Widget DFL-DAG.
- Run measured storage/parallelism experiments.
- Generate Widget-DPM artifact and compare against measurements.
- Replace Montage in the paper once SRAsearch results are internally
  consistent.

### DaYu End-to-End Validation

Status: todo.

Need:

- One small HDF5 workflow with DaYu VOL/VFD profiling.
- Confirm profiler build works in the target environment.
- Confirm `run_dayu_profiling` produces expected stat files.
- Confirm `dayu_vol_sankey`, `dayu_vfd_sankey`, and
  `dayu_combined_sankey` generate valid outputs.
- Upgrade `verification.py` from stub/light DaYu checks to schema/semantic
  validation.

### Integrate with Hasan Agentic-Reasoning Repo

Status: todo.

Repo:

```text
https://github.com/candiceT233/agentic-workflow-orchestrator
```

Need:

- Add Widget-DPM artifact consumption path.
- Add agent instructions for using `dump_candidate_plan_dpm`.
- Add examples for no-DPM, old-DPM, Widget-DPM, and possibly DPM-MT modes.
- Add Jesun and Nathan to the agentic-reasoning/orchestrator repo access as
  requested.

### Per-File Storage Selection

Status: future design/todo.

Current DPM selection unit:

```text
producer application -> consumer application
```

Desired future selection unit:

```text
producer application -> file or file class -> consumer application
```

Reason:

A producer can output different file types with different I/O characteristics.
Pair-level selection can miss cases where one output class should go to shared
storage while another should use local storage or a different parallelism.

Needed design work:

- Classify output files by I/O behavior, size, access pattern, fan-out, and
  lifetime.
- Predict DPM per file or file class.
- Add optimizer constraints to avoid impractical storage fragmentation.
- Decide how HDF5 DaYu dataset-level information maps to physical file
  placement.
- Add reporting that explains file-level choices without overwhelming users.

### Direct Data Flow vs. Intermediate Copy

Status: open question / todo.

Question:

What is the status of DPM for direct data flow versus data flow with an
intermediate copy? Example: the last stage of 1000Genome where intermediate file
plus copy can be faster than direct data flow to parallel storage.

Current short answer:

Widget-DPM can represent and score explicit data movement/copy costs when the
workflow/DPM preparation inserts movement edges and the benchmark profile has
the relevant movement operation. It can also mark unsupported movement profiles
as not applicable. However, the current baseline selection logic is still mainly
candidate-plan scoring for storage pairs and movement cost, not a full planner
that chooses between two alternative DAG structures:

```text
direct write/read on final storage
```

versus

```text
write/read on intermediate local storage -> explicit copy -> final storage
```

What is needed:

- Make "direct path" and "intermediate-copy path" explicit alternative plan
  templates.
- Score both plans with the same DPM components:
  - producer write time;
  - consumer read time;
  - copy/movement time;
  - any serialization or stage-boundary constraints.
- Add selection/reporting logic that can say when intermediate local I/O plus
  copy is predicted faster than direct shared-storage I/O.
- Validate this on the 1000Genome final-stage case before making a paper claim.

Interpretation for now:

Widget-DPM has the ingredients for this comparison, but the explicit
alternative-plan selector and validation are still todo.
